text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,2493 @@
+currently proposed type ascription operator syntax is that it clashes
+with field punning expressions and patterns. That is, if you say:
+`MyStruct { field: Type }`, this is ambiguous with `MyStruct { field: binding }`.
+
+Having said this, there are 3 chief ways to deal with this while retaining
+`:` as a syntax:
+
+1. Accept it and move on. This part of the language grammar will be somewhat
+   unergonomic, but consistency and avoiding ad-hoc syntax is more important.
+   This the proposed solution in this RFC.
+
+2. Accept `MyStruct { field: Type }` where `Type` couldn't be a pattern.
+   Examples of this include `MyStruct { field: Vec<Foo> }`.
+   However, this is an ad-hoc syntax that is likely brittle.
+
+3. Invent some ad-hoc disambiguation syntax. For example, we could entertain
+   the syntax `MyStruct { (field: Type) }` which never parses today.
+   While this could be made to work technically, it does not seem to carry
+   its weight since we expect `MyStruct { field: Type }` to be somewhat rare.
+
+## [RFC 1685] and deprecation schedule
+
+[rust-lang/rust#48309]: https://github.com/rust-lang/rust/pull/48309
+[48309_noted]: https://github.com/rust-lang/rust/pull/48309#issuecomment-391288075
+
+It has been [noted][48309_noted] on [rust-lang/rust#48309] that:
+
+> @scottmcm I think if we had a concrete motivation to be stricter, we could be.
+
+The expedited schedule of making
+
+```rust
+trait Foo {
+    fn bar(Type) { ... }
+}
+```
+
+a warning in 2015 and a hard error in 2018 is motivated by wanting
+to emphasize the nature of function parameters as patterns uniformly.
+For the purposes of improving learnability, we believe it is prudent
+to avoid syntactic inconsistencies in the language.
+
+Furthermore, it was also noted on the same PR by @nikomatsakis that:
+
+> [..] we don't think there's much use of this in the wild,
+> and we have an automated fix.
+
+Thus, we believe it is reasonable to expedite the schedule.
+
+## Precedence of the operator
+
+As explained prior, we change the precedence of `:` when in an expression
+context such that `x : T.foo()` is interpreted as `(x : T).foo()`.
+This precedence change allow users to write readable code when they
+have several method calls by using line-separation such as with:
+
+```rust
+let x = (0..10)
+    .map(some_computation)
+    .collect() : Foo
+    .unwrap()
+    .map(other_computation) : Bar
+    .into() : Baz;
+```
+
+However, if you write this on a single line, or simply consider `x : T.foo()`
+a user might parse this as `x : (T.foo())` instead.
+While at this stage Rust does not support "type-level methods"
+(meaning that this parse currently makes no sense),
+a user may nonetheless make this mistake.
+
+That said, it is still possible for the user to explicitly disambiguate with
+`(x : T).foo()` wherefore this may not become a problem in practice.
+The formatting tool `rustfmt` may also apply such stylings automatically.
+It is important th
```

Comment:
I think we can resolve this using a more general principle that currently holds for `let`s:

> If something is inferred as a particular type when not ascribed, the type can be ascribed explicitly without changing behaviour

If your match has an arm `Foo(x) => ...` and it's giving you `x` as a `&i32`, it feels unacceptable for `Foo(x: &i32)` to suddenly be an error.  It should always be possible to make the inferred type explicit instead.